warp10/src/main/java/io/warp10/plugins/http/WarpScriptHandler.java

```diff
@@ -103,14 +112,29 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
       }
       params.put("headers", headers);
 
-      // Get the payload if the content-type is not application/x-www-form-urlencoded or we do not want to parse the payload
-      if (!MimeTypes.Type.FORM_ENCODED.is(request.getContentType()) || !plugin.isParsePayload(prefix)) {
-        byte[] payload = IOUtils.toByteArray(request.getInputStream());
-        if (0 < payload.length) {
-          params.put("payload", payload);
+      // We have to get the input stream before getting the parameters, else we won't be able to read it!
+      InputStream inputStream = null;
+
+      // The input stream is handled very differently if the WarpScript is expecting a splitted stream or not.
+      if (null == plugin.streamDelimiter(prefix)) {
+        // Get the payload if the content-type is not application/x-www-form-urlencoded or we do not want to parse the payload
+        if ((!MimeTypes.Type.FORM_ENCODED.is(request.getContentType()) || !plugin.isParsePayload(prefix))) {
+          // Ready all the stream and store the payload.
+          byte[] payload = IOUtils.toByteArray(request.getInputStream());
+          if (0 < payload.length) {
+            params.put("payload", payload);
+          }
+        }
+      } else {
+        // The input stream will be splitted and given to the macro later.
+        if (null != request.getHeader("Content-Type") && "application/gzip".equals(request.getHeader("Content-Type"))) {
+          inputStream = new GZIPInputStream(request.getInputStream());
+        } else {
+          inputStream = request.getInputStream();
         }
       }
 
+      // Get the parameters in the URL and payload if we want to parse the x-www-form-urlencoded payload.
```

Comment:
The lines below are unclear, they are executed even if isParsePayload is false, the fact that it works should be described in a comment.